pkg/controller/machineset.go

```diff
@@ -677,6 +678,64 @@ func (c *controller) terminateMachines(inactiveMachines []*v1alpha1.Machine, mac
 	return nil
 }
 
+func (c *controller) terminateStaleMachines(activeMachines, inactiveMachines []*v1alpha1.Machine, machineSet *v1alpha1.MachineSet) error {
+	var (
+		wg                      sync.WaitGroup
+		numOfMachinesToDelete   int
+		numOfRunningMachines    int
+		numOfMachineSetReplicas = int(machineSet.Spec.Replicas)
+		numOfActiveMachines     = len(activeMachines)
+		numOfInactiveMachines   = len(inactiveMachines)
+		errCh                   = make(chan error, numOfInactiveMachines)
+	)
+	defer close(errCh)
+
+	if numOfInactiveMachines == 0 {
+		return nil
+	}
+
+	for _, activeMachine := range activeMachines {
+		if IsMachineRunning(activeMachine) {
+			numOfRunningMachines++
+		}
+	}
+
+	if numOfRunningMachines == 0 || numOfMachineSetReplicas == 0 || c.safetyOptions.FailedMachineDeletionRatio == 1 || c.safetyOptions.FailedMachineDeletionRatio == 0 {
```

Comment:
I was hoping, with `FailedMachineDeletionRatio == 0`, we could simply stop replacing the failed machines.

Reply:
The provided ration will now be taken into account if it equates to `0`.